docs/sphinx_extensions.py

```diff
@@ -160,10 +200,145 @@ def _parse_line_range(self, line_range_str: str, total_lines: int) -> List[int]:
+
+        if "setup-lines" in self.options:
+            setup_line_numbers = parselinenos(self.options["setup-lines"], total_lines)
+            setup_code = "".join([code_lines[i] for i in setup_line_numbers])
+
+        main_line_numbers = parselinenos(self.options["lines"], total_lines)
+        main_code = "".join([code_lines[i] for i in main_line_numbers])
+
+        expected_error = self.options.get("expect-error")
+
+        code_node = nodes.literal_block(main_code, main_code)
+        code_node["language"] = self.options.get("language", "python")
+
+        main_code_lines = main_code.rstrip().split("\n")
+        while main_code_lines and not main_code_lines[-1].strip():
+            main_code_lines.pop()
+        if main_code_lines:
+            last_line = main_code_lines.pop()
+            code_before_last_line = "\n".join(main_code_lines)
+        else:
+            last_line = ""
+            code_before_last_line = ""
+
+        output_io = io.StringIO()
+        exec_globals = {}
+
+        try:
+            with (
+                contextlib.redirect_stdout(output_io),
+                contextlib.redirect_stderr(output_io),
+            ):
+                if setup_code:
+                    exec(setup_code, exec_globals)
+                if code_before_last_line:
+                    exec(code_before_last_line, exec_globals)
+                if last_line:
+                    if self._is_expression(last_line):
+                        result = eval(last_line, exec_globals)
+                        if result is not None:
+                            print(repr(result))  # noqa: T201
+                    else:
+                        exec(last_line, exec_globals)
+
+            # Place this block here: Check if an error was expected but didn't occur
+            error_msg = (
+                f"Expected an error of type '{expected_error}', but no error occurred."
+            )
+            return [code_node, ErrorMessageNode(error_msg)]
+
+        except (KeyboardInterrupt, SystemExit):
+            raise
+        except BaseException as e:
+            expected_error = self.options.get("expect-error")
+
+            if expected_error:
+                if e.__class__.__name__ == expected_error:
+                    # Expected error occurred, show the traceback
+                    error_msg = f"Expected error occurred: {type(e).__name__} - {e}"
+                    return [code_node, ErrorMessageNode(error_msg)]
+
+                error_msg = (
+                    f"Unexpected error type. Expected '{expected_error}', "
+                    f"but got '{e.__class__.__name__}'."
+                )
+                return [code_node, ErrorMessageNode(error_msg)]
+
+            # No expected error provided, show full traceback
+            error_msg = f"Error executing code: {type(e).__name__} - {e}"
+            return [code_node, ErrorMessageNode(error_msg)]
```

Comment:
The expected error handling should just be included in `ExecLiteralInclude`. It does not make sense to create a separate directive for this imo.